Add markup as designed (#574)
unicode-org/message-format-wg: spec/formatting.md

Formatting of a _message_ is defined by the following operations:

> - **_Expression and Markup Resolution_** determines the value of an _expression_ or _markup_,

Reviewer: I think we should make *markup resolution* a separate bullet in this list. Resolving markup is a very different operations from resolving expressions due to markup's not calling any functions, due to the default ignorable property in `formatToString`, and due to the fact that it can only happen inside patterns.

In fact, I'm not even sure if we should call it *resolution*. I guess we still resolve options, but crucially, in its current design, markup is encoded in the data model rather directly.

Author: > I think we should make _markup resolution_ a separate bullet in this list. Resolving markup is a very different operations from resolving expressions due to markup's not calling any functions, due to the default ignorable property in `formatToString`, and due to the fact that it can only happen inside patterns.

I would prefer to keep these together, because the similarities are much greater than the differences.
- Markup resolution does not call external functions, but its results may be defined in the same way as for expressions. We also have expressions such as literals and variables which do not necessarily call any functions.
- Markup resolution uses the same option resolution as functions.
- The behaviour of markup during string formatting is a concern during _formatting_, but not _resolution_.
- Markup resolution only being used for pattern placeholders does not change how it is done.

> In fact, I'm not even sure if we should call it _resolution_. I guess we still resolve options, but crucially, in its current design, markup is encoded in the data model rather directly.

Could you clarify what you mean by "data model" here? The only data model we define in detail is for the message before formatting, from which we need to select, resolve, and format a pattern during formatting. We left out formatted parts, so the output structure is not well defined.

Reviewer 2: My first reaction to the draft was like @stasm's: it's okay to be somewhat repetitious in a spec. Keeping a wall between the two similar-but-unlike concepts might be useful. 

I didn't make that comment in the end, because there is a lot of commonality and the resolution into different buckets does happen later.

> We also have expressions such as literals and variables which do not necessarily call any functions.

Are you sure? If there is a placeholder `{|zebra|}`, I'm pretty sure that invokes the `:string` formatting function. If there is a placeholder `{$foo}`, I'd allow the implementation to introspect `foo`'s type, but I would expect *some* function to be called in the end. We don't want it to be an error to write messages like `Hello {$user}`...

Author: > If there is a placeholder `{|zebra|}`, I'm pretty sure that invokes the `:string` formatting function. If there is a placeholder `{$foo}`, I'd allow the implementation to introspect `foo`'s type, but I would expect _some_ function to be called in the end.

I agree that this is a reasonable thing for an implementation to do, but I don't think we should expect it to _always_ happen. There are also reasonable further questions, such as: What formatter is called for `{|zebra|}` if an implementation allows for overriding default formatters, and `:string` is so overridden? Is it the core `:string` formatter, or the user-provided `:string` formatter that handles the zebra?

I think that's a question we ought to _not_ resolve in this spec, leaving it instead to implementations.

Reviewer 2: > I think that's a question we ought to not resolve in this spec, leaving it instead to implementations.

I agree. Hmm... what concrete change are we proposing with this thread?

In _selectors_, the resolved value of an _expression_ is used for _pattern selection_.

> In a _pattern_, the resolved value of _expressions_ and _markup_ is used in their _formatting_.

Reviewer 2: "their" is incorrect, since it refers to "value". Perhaps a bit of avoidance can be added too:

```suggestion
In a _pattern_, the resolved value of an _expression_ or _markup_ is used in its _formatting_.
```

   If the call fails or does not return a valid value,
   emit a Resolution error and use a _fallback value_ for the _expression_.

#### Option Resolution

The result of resolving _option_ values is a mapping of string identifiers to values.

For each _option_:

- Resolve the _identifier_ of the _option_.
- If the _option_'s _identifier_ already exists in the resolved mapping of _options_,
   emit a Duplicate Option Name error.
- If the _option_'s right-hand side successfully resolves to a value,
   bind the _identifier_ of the _option_ to the resolved value in the mapping.
- Otherwise, bind the _identifier_ of the _option_ to an unresolved value in the mapping.
   Implementations MAY later remove this value before calling the _function_.
   (Note that an Unresolved Variable error will have been emitted.)

Errors MAY be emitted during _option resolution_,
but it always resolves to some mapping of string identifiers to values.
This mapping can be empty.

### Markup Resolution

Unlike _functions_, the resolution of _markup_ is not customizable.

> 

Reviewer 2: This might be too opinionated? If I'm understanding this correctly, this doesn't permit implementations to support plug-in markup handlers that work similar to functions. While that might be the right solution for some implementations, it might be be what everyone expects. 

We do want, I think, to say that markup's fallback is to nothing?

Author: Note that this statement is specifically about _resolution_, and does not limit what may be done afterwards. A "plug-in markup handler" could post-process resolved values into different shapes, and it could even have access to the whole span of parts between open & close elements.

Reviewer 2: What I'm saying is: it's not "afterwards" at all, it might be **now** as part of the resolution, if the implementation chooses to make it work that way.

Instead of saying this (which doesn't really help implementations either way), perhaps state what our limits are:

```suggestion
This specification does not fully define how _markup_ is resolved.
Implementations can handle _markup_ according to their own needs, so long as the following
requirements are met.
```

The resolved value of _markup_ includes the following fields:

- The type of the markup: open, standalone, or close

> - The _name_ of the _markup_

Reviewer 2: identifier?

Author: ```suggestion
- The _identifier_ of the _markup_
```

- For _markup-open_ and _markup_standalone_,
  the resolved _options_ values after _option resolution_.

> The resolution of _markup_ MUST always succeed.

Reviewer 3: This doesn't look quite right for a use of RFC 2119 "MUST".

Author: The usage here is intentional, at least on my part: An implementation must not allow for the resolution of any `{#foo}` part to ever fail.

- A string with associated attributes for portions of its text.
- A flat sequence of objects corresponding to each resolved value.
- A hierarchical structure of objects that group spans of resolved values,
  such as sequences delimited by _markup-open_ and _markup-close_ _placeholders_.

Implementations SHOULD provide _formatting_ result types that match user needs,
including situations that require further processing of formatted messages.
Implementations SHOULD encourage users to consider a formatted localised string
as an opaque data structure, suitable only for presentation.

> When formatting to a string, the default representation of all _markup_
> MUST be an empty string.
> Implementations MAY offer functionality for customizing this,
> such as by emitting XML-ish tags for each _markup_.

Reviewer 2: I would tighten this up a bit. The normative text is appropriate, but I would avoid the MUST. I would also make the example a bit more fleshed out.

```suggestion
The default representation of all types of _markup_ placeholder
is an empty string.
That is, unless otherwise customized, open, close and standalone
markup placeholders are omitted from the output.
Implementations MAY offer functionality to customize this behavior.

> For example, an implementation might replace markup with
> styling attributes in an attributed string API 
> or it might replace the placeholders with HTML markup.
```

Author: Your suggestion leaves out the "When formatting to a string" qualifier. Is that intentional?

Reviewer 2: Yes.
